pkg/systemd/services/services.jsx

```diff
@@ -66,20 +67,36 @@ let monotonic_timer_base = null; // µs
 export const MAX_UINT64 = 2 ** 64 - 1;
 
 export function updateTime() {
-    Promise.all([
-        cockpit.file("/proc/uptime").read(),
-        cockpit.spawn(["date", "+%s"])
-    ])
-            .then(([uptime_contents, date_output]) => {
-                clock_realtime_now = parseInt(date_output, 10) * 1000;
-
-                // first number in /proc/uptime is time since boot (CLOCK_BOOTTIME) in seconds with two fractional digits
-                const uptime = parseFloat(uptime_contents.split(' ')[0]);
-                const clock_boottime_now = parseInt(uptime * 1000000, 10);
-
-                monotonic_timer_base = clock_realtime_now * 1000 - clock_boottime_now;
+    /* To correctly interpret monotonic timers, we need to read CLOCK_MONOTONIC. This cannot be read with shell tools,
+     * and the only source in /proc is /proc/timer_list which is readable by root only (and also rather hard to parse). */
+    python.spawn("import time; print(time.clock_gettime_ns(time.CLOCK_REALTIME), time.clock_gettime_ns(time.CLOCK_MONOTONIC))",
+                 null, { err: "message" })
+            .then(output => {
+                const [realtime, monotonic] = output.split(' ').map(ns => parseInt(ns));
+                clock_realtime_now = realtime / 1000000;
+                monotonic_timer_base = (realtime - monotonic) / 1000;
             })
-            .catch(ex => console.warn("Failed to read boot time:", ex.toString()));
+            .catch(ex => {
+                /* If Python is not available, fall back to reading CLOCK_BOOTTIME from /proc/uptime. This will
+                 * give incorrect results if the machine was suspended during the current boot, but it's the
+                 * best approximation that we can get. */
```

Comment:
Hmm, I would say wrong information is worse than no information. Can we fall back to something horrible that gives the right information, like spawning "systemctl"?

Reply:
It's hard to decide if it's wrong, unfortunately.  I discussed some options in #18418 . 

Most servers don't get suspended, this is mostly a desktop thing -- thus it would be bearable, and at least fix the situation for the 90% case. For the fallback, I could imagine parsing /proc/timer_list with `superuser: require`, and just don't show any timer information at all if you are not root. This would then not even work for the "Users" units, though.

Reply:
A simpler alternative would of course be to just not show any timer information without python. I don't have a stong opinion either way.. Soon, we will require Python anyway :grin:

Reply:
> A simpler alternative would of course be to just not show any timer information without python.

Yeah, personally I think this is the right thing to do.

Reply:
> > A simpler alternative would of course be to just not show any timer information without python.
> 
> Yeah, personally I think this is the right thing to do.

And if people complain, we could somehow export the timers from cockpit-bridge.

Reply:
Ack. Also, it doesn't actually affect *all* timers -- e.g. OnCalendar= is fine. Only monotonic ones. Let's do that then.